src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -0,0 +1,1093 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2013-2015, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ----------------------------------------------------------------------
+ */
+
+ /** @file
+  * Implementation of TemporalMemory
+  */
+
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <iterator>
+#include <vector>
+#include <boost/tuple/tuple.hpp>
+
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include <kj/std/iostream.h>
+
+#include <nupic/algorithms/Connections.hpp>
+#include <nupic/algorithms/TemporalMemory.hpp>
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::algorithms::connections;
+using namespace nupic::algorithms::temporal_memory;
+
+TemporalMemory::TemporalMemory()
+{
+  version_ = 1;
+}
+
+TemporalMemory::TemporalMemory(
+  vector<UInt> columnDimensions,
+  Int cellsPerColumn,
+  Int activationThreshold,
+  Permanence initialPermanence,
+  Permanence connectedPermanence,
+  Int minThreshold,
+  Int maxNewSynapseCount,
+  Permanence permanenceIncrement,
+  Permanence permanenceDecrement,
+  Int seed)
+{
+  initialize(
+    columnDimensions,
+    cellsPerColumn,
+    activationThreshold,
+    initialPermanence,
+    connectedPermanence,
+    minThreshold,
+    maxNewSynapseCount,
+    permanenceIncrement,
+    permanenceDecrement,
+    seed);
+}
+
+TemporalMemory::~TemporalMemory()
+{
+}
+
+void TemporalMemory::initialize(
+  vector<UInt> columnDimensions,
+  Int cellsPerColumn,
+  Int activationThreshold,
+  Permanence initialPermanence,
+  Permanence connectedPermanence,
+  Int minThreshold,
+  Int maxNewSynapseCount,
+  Permanence permanenceIncrement,
+  Permanence permanenceDecrement,
+  Int seed)
+{
+  // Error checking
+  if (columnDimensions.size() <= 0)
+    NTA_THROW << "Number of column dimensions must be greater than 0";
+
+  if (cellsPerColumn <= 0)
+    NTA_THROW << "Number of cells per column must be greater than 0";
+
+  // TODO: Validate all parameters (and add validation tests)
```

Comment:
https://github.com/numenta/nupic.core/issues/504 and a comment about it placed in code below this comment.